File: day5/day5.cpp
#include <iostream>
#include <sstream>
#include <fstream>
#include <string>
#include <vector>

using namespace std;
using u64 = unsigned long long;

struct Mapping {
    Mapping() {
        this->dst = 0;
        this->src = 0;
        this->length = 0;
    }
    Mapping(u64 dst, u64 src, u64 length) {
        this->dst = dst;
        this->src = src;
        this->length = length;
    }
    u64 dst;
    u64 src;
    u64 length;
};

template <typename T>
void show(vector<T> vec) {
    for (const T& item : vec) {
        cout << item << " ";
    }
    printf("\n");
}


vector<string> split(const string &s, char delimeter) {
    vector<string> result;
    stringstream ss(s);
    string item;

    while (getline(ss, item, delimeter)) {
        if (item.length() <= 0) {
            continue;
        } 
        result.push_back(item);
    }

    return result;
} 

vector<u64> initSeeds(string line) {
    string rhs = split(line, ':')[1];
    vector<string> seedStrs = split(rhs, ' ');

    vector<u64> seeds;
    u64 seedStart = 0;

    for (size_t i = 0; i < seedStrs.size(); i++) {
        string seedItem = seedStrs[i];
        u64 seedNum = stoull(seedItem);

        if (i % 2 == 0) {
             seedStart = seedNum;
        } else {
            for (u64 j = seedStart; j < seedStart + seedNum; j++) {
                seeds.push_back(j);
            }
        }
    }

    return seeds;
}

void initMap(vector<Mapping*> &mapping, string line) {
    vector<string> mappingStrs = split(line, ' ');
    vector<u64> mappingNums;

    for (const string& str : mappingStrs) {
        u64 num = stoull(str);
        mappingNums.push_back(num);
    }

    Mapping *newMapping = new Mapping(mappingNums[0], mappingNums[1], mappingNums[2]);
    mapping.push_back(newMapping);
}

u64 findCorrespondingValue(u64 key, vector<Mapping*> &mappings) {
    for (const Mapping* mapping : mappings) {
        u64 src = mapping->src;
        u64 end = src + mapping->length;

        if (key >= src && key < end) {
            u64 diff = key - src;
            return mapping->dst + diff;
        }
    }

    return key;
}

void clearMappings(vector<Mapping*>& currMappings) {
    for (const Mapping* mapping : currMappings) {
        delete mapping;
    }
    currMappings.clear();
}

int main() {
    ifstream inputFile("d5");
    string line;
    vector<string> buffer;
    while (getline(inputFile, line)) {
        buffer.push_back(line);
    }

    vector<u64> currKeys = initSeeds(buffer[0]);
    vector<Mapping*> currMappings;

    u64 lineNumber = 3;

    u64 translations = 7;
    while (translations > 0) {
        while (true) {
            line = buffer[lineNumber];
            if (line.size() == 0) {
                lineNumber += 2;
                break;
            }
            initMap(currMappings, line);
            lineNumber++;
        }

        for (size_t i = 0; i < currKeys.size(); i++) {
            currKeys[i] = findCorrespondingValue(currKeys[i], currMappings);
        }
        clearMappings(currMappings);
        printf("translations %llu\n", translations);
        translations--;
    }

    u64 minLocation = currKeys[0];
    for (size_t i = 1; i < currKeys.size(); i++) {
        if (currKeys[i] < minLocation) {
            minLocation = currKeys[i];
        }
    }
    printf("Min Location: %llu\n", minLocation);
}

File: day5/day5.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

type Mapping struct {
    dst int
    src int
    length int
}

func newMappings(dst int, src int, length int) Mapping {
    return Mapping{
        dst: dst,
        src: src,
        length: length,
    }
}

func initSeeds(line string) []int {
    rhs := strings.Split(line, ": ")[1]
    seedStrs := strings.Split(rhs, " ")

    var seeds []int
    for _, seed := range(seedStrs) {
        seedNum, err := strconv.Atoi(seed)
        if err != nil {
            continue
        }
        seeds = append(seeds, seedNum)
    }

    return seeds
}

func initSeedsPart2(line string) []int {
    rhs := strings.Split(line, ": ")[1]
    seedStrs := strings.Split(rhs, " ")

    var seeds []int
    var seedStart int
    for i, seed := range(seedStrs) {
        if i % 2 == 0 {
            seedNum, err := strconv.Atoi(seed)
            if err != nil {
                panic("NaN")
            }
            seedStart = seedNum
        } else {
            seedRange, err := strconv.Atoi(seed)
            if err != nil {
                panic("NaN")
            }
            for i := seedStart; i < seedStart + seedRange; i++ {
                seeds = append(seeds, i)
            }
        }
    }

    return seeds
}

func initMap(mapping *[]Mapping, line string) {
    mappingStrs := strings.Split(line, " ")
    var mappingNums []int

    for _, str := range(mappingStrs) {
        num, err := strconv.Atoi(str)
        if err != nil {
            continue
        }
        mappingNums = append(mappingNums, num)
    }

    newMapping := newMappings(mappingNums[0], mappingNums[1], mappingNums[2])
    *mapping = append(*mapping, newMapping)
}

func findCorrespondingValue(key int, mappings []Mapping) int {
    for _, mapping := range(mappings) {
        srcStart := mapping.src
        srcEnd := srcStart + mapping.length

        if key >= srcStart && key < srcEnd {
            diff := key - srcStart
            return mapping.dst + diff
        }
    }

    return key
}

func main() {
    filePath := "d5"
    file, _ := os.Open(filePath)
    defer file.Close()

    scanner := bufio.NewScanner(file)
    var lines []string

    for scanner.Scan() {
        line := scanner.Text()
        lines = append(lines, line)
    }

    var seeds                   []int
    var seedToSoil              []Mapping
    var soilToFertilizer        []Mapping
    var fertilizerToWater       []Mapping
    var waterToLight            []Mapping
    var lightToTemperature      []Mapping
    var temperatureToHumidity   []Mapping
    var humidityToLocation      []Mapping

    section := 0
    onInfoLine := false

    for _, line := range(lines) {
        if len(line) == 0 {
            section += 1
            onInfoLine = true
            continue
        }
        switch section {
        case 0:
            seeds = initSeedsPart2(line) 
        case 1:
            if onInfoLine {
                onInfoLine = false
                continue
            } 
            initMap(&seedToSoil, line) 
        case 2: 
            if onInfoLine {
                onInfoLine = false
                continue
            } 
            initMap(&soilToFertilizer, line)
        case 3:
            if onInfoLine {
                onInfoLine = false
                continue
            }
            initMap(&fertilizerToWater, line)
        case 4:
            if onInfoLine {
                onInfoLine = false
                continue
            }
            initMap(&waterToLight, line)
        case 5:
            if onInfoLine {
                onInfoLine = false
                continue
            }
            initMap(&lightToTemperature, line)
        case 6:
            if onInfoLine {
                onInfoLine = false
                continue
            }
            initMap(&temperatureToHumidity, line) 
        case 7:
            if onInfoLine {
                onInfoLine = false
                continue
            }
            initMap(&humidityToLocation, line)
        default:
            break
        }
    }

    // fmt.Println(seeds)
    // fmt.Println(seedToSoil)
    // fmt.Println(soilToFertilizer)
    // fmt.Println(fertilizerToWater)
    // fmt.Println(waterToLight)
    // fmt.Println(lightToTemperature)
    // fmt.Println(temperatureToHumidity)
    // fmt.Println(humidityToLocation)

    var soils []int
    var fertilizers []int
    var waters []int
    var lights []int
    var temperatures []int
    var humidities []int
    var locations []int

    for _, seed := range(seeds) {
        soil := findCorrespondingValue(seed, seedToSoil)
        soils = append(soils, soil)
    }
    // fmt.Println(soils)
    for _, soil := range(soils) {
        fertilizer := findCorrespondingValue(soil, soilToFertilizer)
        fertilizers = append(fertilizers, fertilizer)
    }
    // fmt.Println(fertilizers)
    for _, fertilizer := range(fertilizers) {
        water := findCorrespondingValue(fertilizer, fertilizerToWater)
        waters = append(waters, water)
    }
    // fmt.Println(waters)
    for _, water := range(waters) {
        light := findCorrespondingValue(water, waterToLight)
        lights = append(lights, light)
    } 
    // fmt.Println(lights)
    for _, light := range(lights) {
        temperature := findCorrespondingValue(light, lightToTemperature)
        temperatures = append(temperatures, temperature)
    }
    // fmt.Println(temperatures)
    for _, temperature := range(temperatures) {
        humidity := findCorrespondingValue(temperature, temperatureToHumidity)
        humidities = append(humidities, humidity)
    }
    // fmt.Println(humidities)
    for _, humidity := range(humidities) {
        location := findCorrespondingValue(humidity, humidityToLocation)
        locations = append(locations, location)
    }
    // fmt.Println(locations)

    fmt.Println(slices.Min(locations))
}
